src/psyclone/psyir/nodes/call.py

```diff
@@ -294,6 +296,44 @@ def _validate_child(position, child):
         '''
         return isinstance(child, DataNode)
 
+    def reference_accesses(self, var_accesses):
+        '''
+        Updates the supplied var_accesses object with information on the
+        arguments passed to this call.
+
+        TODO #446 - all arguments that are passed by reference are currently
+        marked as having READWRITE access (unless we know that the routine is
+        PURE). We could do better than this if we have the PSyIR of the called
+        Routine.
+
+        :param var_accesses: VariablesAccessInfo instance that stores the
+            information about variable accesses.
+        :type var_accesses:
+            :py:class:`psyclone.core.VariablesAccessInfo`
+
+        '''
+        if self.is_pure:
+            # If the called routine is pure then any arguments are only
+            # read.
+            default_access = AccessType.READ
+        else:
+            # We conservatively default to READWRITE otherwise (TODO #446).
+            default_access = AccessType.READWRITE
+
+        for arg in self.children:
+            if isinstance(arg, Reference):
+                # This argument is pass-by-reference.
+                sig, indices_list = arg.get_signature_and_indices()
+                var_accesses.add_access(sig, default_access, arg)
+                # Any symbols referenced in any index expressions are READ.
+                for indices in indices_list:
+                    for idx in indices:
+                        idx.reference_accesses(var_accesses)
+            else:
+                # This argument is ultimately pass-by-value so any
+                # inputs to it are READ.
+                arg.reference_accesses(var_accesses)
```

Comment:
I was almost ready to approve the PR but this comment is wrong, even though the implementation is right.

An expression.reference_access will decide its access type, it is not necessarily READ. For instance:
`call my_pure_sub(a, b, my_impure_increment_and_return_func(c))` would make a:READ, b:READ, c: READWRITE

Maybe you can add this nested example in your `test_call_reference_accesses`

Reply:
I think this is already covered by:

https://github.com/stfc/PSyclone/blob/2b6e8dfc90608a560fb3b33efc42e8ce8c43f2cb/src/psyclone/tests/psyir/nodes/call_test.py#L423-L430

Reply:
I've updated the Comment now though (well spotted).